spannerlib/wrappers/spannerlib-python/spannerlib-python/google/cloud/spannerlib/connection.py

```diff
@@ -69,3 +79,108 @@ def _close_lib_object(self) -> None:
+
+        logger.info(
+            "Executing SQL on connection ID: %d for pool ID: %d",
+            self.oid,
+            self.pool.oid,
+        )
+
+        request_bytes = ExecuteSqlRequest.serialize(request)
+
+        # Call the Go library function to execute the SQL statement.
+        with self.spannerlib.execute(
+            self.pool.oid, self.oid, request_bytes
+        ) as msg:
+            msg.raise_if_error()
+            logger.info(
+                "SQL execution successful on connection ID: %d."
+                "Got Rows ID: %d",
+                self.oid,
+                msg.object_id,
+            )
+            return Rows(msg.object_id, self.pool, self)
+
+    def execute_batch(
+        self, request: ExecuteBatchDmlRequest
+    ) -> ExecuteBatchDmlResponse:
+        """Executes a batch of DML statements on the connection.
+
+        Args:
+            request: The ExecuteBatchDmlRequest object.
+
+        Returns:
+            An ExecuteBatchDmlResponse object representing the result
+            of the execution.
+        """
+        if self.closed:
+            raise RuntimeError("Connection is closed.")
+
+        logger.info(
+            "Executing batch DML on connection ID: %d for pool ID: %d",
+            self.oid,
+            self.pool.oid,
+        )
+
+        request_bytes = ExecuteBatchDmlRequest.serialize(request)
+
+        # Call the Go library function to execute the batch DML statement.
+        with self.spannerlib.execute_batch(
+            self.pool.oid,
+            self.oid,
+            request_bytes,
+        ) as msg:
+            msg.raise_if_error()
+            logger.info(
+                "Batch DML execution successful on connection ID: %d.",
+                self.oid,
+            )
+            response_bytes = to_bytes(msg.msg, msg.msg_len)
+            return ExecuteBatchDmlResponse.deserialize(response_bytes)
+
+    def write_mutations(
+        self, request: BatchWriteRequest.MutationGroup
+    ) -> CommitResponse:
+        """Writes a mutation to the connection.
+
+        Args:
+            request: The BatchWriteRequest_MutationGroup object.
+
+        Returns:
+            A CommitResponse object.
+        """
+        if self.closed:
+            raise RuntimeError("Connection is closed.")
+
+        logger.info(
+            "Writing mutation on connection ID: %d for pool ID: %d",
+            self.oid,
+            self.pool.oid,
+        )
+
+        request_bytes = BatchWriteRequest.MutationGroup.serialize(request)
+
+        # Call the Go library function to write the mutation.
+        with self.spannerlib.write_mutations(
+            self.pool.oid,
+            self.oid,
+            request_bytes,
+        ) as msg:
+            msg.raise_if_error()
+            logger.info(
+                "Mutation write successful on connection ID: %d.", self.oid
+            )
+            response_bytes = to_bytes(msg.msg, msg.msg_len)
```

Comment:
This needs a check whether a message was actually returned. The `WriteMutations` method is a bit tricky:
1. It returns a `CommitResponse` if the mutations were written directly (that is: The connection did not have an active transaction)
2. It returns nothing if the mutations were only buffered in the current transaction.